contracts/utils/math/Math.sol

```diff
@@ -335,6 +335,63 @@ library Math {
         }
     }
 
+    /**
+     * @dev Variant of {modExp} that supports inputs of arbitrary length.
+     */
+    function modExp(bytes memory b, bytes memory e, bytes memory m) internal view returns (bytes memory) {
+        (bool success, bytes memory result) = tryModExp(b, e, m);
+        if (!success) {
+            if (_zeroBytes(m)) {
+                Panic.panic(Panic.DIVISION_BY_ZERO);
+            } else {
+                revert Address.FailedInnerCall();
+            }
+        }
+        return result;
+    }
+
+    /**
+     * @dev Variant of {tryModExp} that supports inputs of arbitrary length.
+     */
+    function tryModExp(
+        bytes memory b,
+        bytes memory e,
+        bytes memory m
+    ) internal view returns (bool success, bytes memory result) {
+        if (_zeroBytes(m)) return (false, result);
+
+        uint256 mLen = m.length;
+
+        // Encode call args in result and move the free memory pointer
+        result = abi.encodePacked(b.length, e.length, mLen, b, e, m);
+
+        /// @solidity memory-safe-assembly
+        assembly {
+            // Write result on top of args to avoid allocating extra memory.
+            // | Offset    | Content      | Content (Hex)                                                      |
+            // |-----------|--------------|--------------------------------------------------------------------|
+            // | 0x00:0x1f | args length  | 0x<.......................................20+20+20+bLen+eLen+mLen> |
+            // | 0x20+mLen | result       | 0x<........................................................result> |
+            // | 0x..:0x.. | dirty bytes  | 0x<............................................20+20+20+bLen+eLen> |
+            success := staticcall(gas(), 0x05, add(result, 0x20), mload(result), add(result, 0x20), mLen)
+            // Overwrite the length.
+            // result.length > returndatasize() is guaranteed because returndatasize() == m.length
+            mstore(result, mLen)
+        }
```

Comment:
This still leaks:

`result = abi.encodePacked(b.length, e.length, mLen, b, e, m);`  allocates a memory array that is 
- 0x20 for the length (of the bytes being encoded)
- 0x20 for the b.length
- 0x20 for the e.length
- 0x20 for the m.length
- 0x20 * b.length for b
- 0x20 * e.length for e
- 0x20 * m.length for m

For a total of `0x80 + 0x20 * (b.length + e.length + m.length)`. That is by how much the free memory pointer is moved.

For result, we only need:
- 0x20 for the length of result (that is value of m.length)
- 0x20 * m.length for the result.

So the current version moves the free memory ptr by `0x60 + 0x20 * (b.length + e.length)` more than needed. That is memory being allocated for no usage.

Reply:
Yes, that's right. But I'm unsure of allocating the pointer after `m`. All the other elements haven't been zeroed out. That's a dangerous side effect.

Reply:
> All the other elements haven't been zeroed out. That's a dangerous side effect.

IMO that is not dangerous. Its clear that there no one should expect the bytes after the free memory ptr to be zeroed. 

See the warning [here](https://docs.soliditylang.org/en/latest/internals/layout_in_memory.html)
![Capture d’écran du 2024-02-13 15-53-03](https://github.com/OpenZeppelin/openzeppelin-contracts/assets/2432299/0349faa5-e3da-48b9-8c84-7f4ca60e5a3f)

A typical example is [this function](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/Clones.sol#L68-L84). We write a lot of stuff in the free memory space, use that to do some computations, and never clean it. The space remains unallocated and is not zero.

Reply:
> A typical example is [this function](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/Clones.sol#L68-L84). We write a lot of stuff in the free memory space, use that to do some computations, and never clean it. The space remains unallocated and is not zero.

Right, this example convinced me. My main worry is that the modexp implementation may come as a report through Immunefi and I would've taken it as valid. The Clones example is evidence this is not an issue.

Note that I don't agree that developers should always know the default Solidity behavior. Imo they should know if they explicitly decide to deal with a particular feature that has side effects, but here the side effects are not obvious for regular language users.

Solving.